### PR TITLE
[NON-BREAKING] ES2015 Proxy approach (requires Node.js 6)

### PR DESCRIPTION
**Pull request by schnittstabil**
See: #29

### CODE REVIEW

**Comment by schnittstabil** [comment]
> `argumentList` can just be passed as an array.

If I understand you correctly, that would also imply:
```js
const processFn = (fn, opts) => function (args) {
// …
}
// …
		get: (target, key) => {
			// …
			return function() {
				cached(arguments); // or array version of arguments
			};
		}
```

**Comment by sindresorhus** [comment]
Not entirely sure what your example there is supposed to imply?

File: index.js [hunk 1]
```diff
@@ -14,11 +14,7 @@ const processFn = (fn, opts) => function () {
 			if (err) {
 				reject(err);
 			} else if (opts.multiArgs) {
-				const results = new Array(arguments.length - 1);
-
-				for (let i = 1; i < arguments.length; i++) {
-					results[i - 1] = arguments[i];
-				}
+				const [, ...results] = arguments;
```
Comment (sindresorhus) [inline index.js]: No need to use `arguments`. Just use the `rest` operator in the function. Can switch to arrow function then too.

Comment (schnittstabil) [inline index.js]: Sadly, arrow functions don't have `arguments`...

Comment (schnittstabil) [inline index.js]: Ah, ok....

Comment (sindresorhus) [inline index.js]: Re-read my comment. `rest` parameters ftw ;)

**Comment by schnittstabil** [comment]
I don't see benefits, e.g every `get` returns a newly created `function`:
```js
            return function() {
                return cached(arguments); // or array version of arguments
            };
```

File: index.js (hunk 1 above)
Comment (schnittstabil) [inline index.js]: @sindresorhus Which version do you prefere:
```js
		args.push((err, result, ...results) => {
			if (err) {
				reject(err);
			} else if (opts.multiArgs) {
				resolve([result, ...results]);
			} else {
				resolve(result);
			}
		});
```
or
```js
		args.push((err, ...results) => {
			if (err) {
				reject(err);
			} else if (opts.multiArgs) {
				resolve(results);
			} else {
				resolve(results[0]);
			}
		});
```

**Comment by sindresorhus** [comment]
Easier to explain with code. This is what I was thinking: https://github.com/sindresorhus/pify/commit/65d5fc9b14e7778871b189d65c07316273014c0c

**Comment by schnittstabil** [comment]
https://github.com/sindresorhus/pify/commit/65d5fc9b14e7778871b189d65c07316273014c0c: Maybe it's not covered by the tests, but I think you have to change the return value of `get` too!?

File: index.js (hunk 1 above)
Comment (sindresorhus) [inline index.js]: First. We should optimize for the common case.

**Comment by sindresorhus** [comment]
@schnittstabil I wanna ship this, but can't realistically *just* target Node.js 6 for a good while. How about we put this in a separate file and ship it now? People targeting Node.js 6 can use it, and it wouldn't make a difference for Node.js 4 users. For example: `const pify = require('pify/proxy');` (Or a better name?)

**Comment by schnittstabil** [comment]
`require('pify/proxy')` sounds reasonable.

Honestly, I do not believe it works to all extend, but it seems ready for most use cases. Hence, I would suggest to either release it for internal usage only or document it as _beta_ or similar.

**Comment by sindresorhus** [comment]
Could you move it to a separate file?

**Comment by schnittstabil** [comment]
Sorry, not until Monday.

**Comment by schnittstabil** [comment]
@sindresorhus Doesn't really LGTM, but proxy stuff moved into `proxy.js`.

Remaining issues I see so far:
1. How should we deal with `[[Set]]`? For example:
    ```js
    const fsP = pify(fs);
    fsP.stat = p => new Promise(…);
    // and
    fsP.stat = (p, cb) => {…};
    ```
2. How should we deal with non-writable, non-configurable properties? For example:
    ```js
    const fsP = pify(fs);
    Object.freeze(fs);
    fs.stat === fsP.stat // <- MUST be true!?
    ```
    From [[[Get]] (P, Receiver)](http://www.ecma-international.org/ecma-262/6.0/#sec-proxy-object-internal-methods-and-internal-slots-get-p-receiver):
    > [[Get]] for proxy objects enforces the following invariants:
    > * The value reported for a property must be the same as the value of the corresponding target object  property if the target object property is a non-writable, non-configurable own data property.

File: test-proxy.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
+import fs from 'fs';
```
Comment (schnittstabil) [inline test-proxy.js]: `test-proxy.js` is an one-to-one duplicate of `test.js` :unamused:

File: test-proxy.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+import test from 'ava';
+import pinkiePromise from 'pinkie-promise';
+import fn from './proxy';
```
Comment (schnittstabil) [inline test-proxy.js]: `test-proxy.js` is (almost) an one-to-one duplicate of `test.js` :unamused:

**Comment by oncletom** [comment]
You don't want to have two branches, and to publish to two different npm tags?

The breaking change would occur when you decide to move the `node6-proxy` to `latest`.

Just a thought, not even an opinion :-)

**Comment by schnittstabil** [comment]
@oncletom I'm not sure if I understand you correctly. As far as I can see, merging this PR into `master` is not a breaking change (anymore):

* nothing changed for node4 and node6 users if they `require('pify')`
* node6 users _can_ also `require('pify/proxy')`, thus proxies are only an _additional functionality_

**Comment by oncletom** [comment]
@schnittstabil yeah sorry, I was not clear at all. I rephrased some of my thoughts and I totally get the fact you ship both approaches in the same branch, but with an additional entrypoint for people having a node environment which has `Proxy` enabled.

**Comment by sindresorhus** [comment]
1. Just let them override. We shouldn't do anything. Pretty edge-casy anyways.

2. We document it as a caveat. I really doubt anyone will hit this. I've yet to see `Object.freeze()` actually used. And if they try using it, it will throw:

> 'get' on proxy: property 'stat' is a read-only and non-configurable data property on the proxy target but the proxy did not return its actual value

File: test-proxy.js (hunk 3 above)
Comment (sindresorhus) [inline test-proxy.js]: That's ok for now. We'll either replace proxy with the current approach or adopt both in the same API once we can target Node.js 6.

File: proxy.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
+'use strict';
+
+const processFn = (fn, opts) => function () {
+	const that = this;
+	const P = opts.promiseModule;
+	const args = new Array(arguments.length);
+
+	for (let i = 0; i < arguments.length; i++) {
+		args[i] = arguments[i];
+	}
```
Comment (sindresorhus) [inline proxy.js]: Could you also integrate https://github.com/sindresorhus/pify/commit/65d5fc9b14e7778871b189d65c07316273014c0c? Would like to remove this boilerplate code.

Comment (schnittstabil) [inline proxy.js]: Unfortunately, the version below breaks optimization:

```js
const processFn = (fn, opts) => function (...args) { …
```

---

And the following version don't work well too:

```js
const processFn = (fn, opts) => function (args) { …
```

`handler.get` have to be able to return `processFn(x, opts)` with the old function signature. We could fix the signature in `handler.get`, but that wouldn't be very sensible in my opinion.

Comment (sindresorhus) [inline proxy.js]: Alright, let's just hold off on changing that until Node.js 8 which should fix the perf deoptimization for that.

**Comment by sindresorhus** [comment]
@schnittstabil I totally forgot about this one. Anything left before we can merge? Can you fix the merge conflict?

**Pull request closed without merging**